### ISSUE TITLE
Support BOMs/dependencyManagement

### ISSUE THREAD

**Issue by mandrean** [coursier/coursier#1390]
Correct me if I'm wrong, but to me it seems that the Coursier CLI doesn't support fetching/resolving the `dependencyManagement` dependencies in BOMs?

For example: https://github.com/FasterXML/jackson-bom/blob/master/pom.xml#L75

I've tried all kinds of variants
```
$ coursier resolve com.fasterxml.jackson:jackson-bom:2.9.10 -t
$ coursier fetch com.fasterxml.jackson:jackson-bom:2.9.10 -j dep-tree.json && cat dep-tree.json
```

and more.

Also see https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Management

>The dependency management section is a mechanism for centralizing dependency information. When you have a set of projects that inherits a common parent it's possible to put all information about the dependency in the common POM and have simpler references to the artifacts in the child POMs.

Commonly used in large companies with a lot of teams developing various services in order to align them on external dependency versions!

**Comment by alexarchambault** [comment]
You mean, you'd like the CLI to fetch all the dependencies of the `dependencyManagement` section? It's not something that's supported as is. What would be the use case for that?

Resolution does take into account parent POMs / import scope during resolution though.

**Comment by mandrean** [comment]
Partially, the use-case is: https://github.com/bazelbuild/rules_jvm_external/issues/229

But also more broadly speaking:

Many companies (including Spotify, where I'm at) use BOMs to centrally manage versions of various company-wide internal dependencies. Also some open-source projects have support for that, like: https://github.com/FasterXML/jackson-bom

As it is today, if you run `coursier fetch` on some artifact that happens to rely on a BOM to resolve some of its dependencies, then `coursier`, would make an incomplete job at downloading all the transitive dependencies  (`mvn` can do this, however).

Same if you use `--json-output-file` - the dependency resolution list there is in fact incomplete!

**Comment by kohsuke** [comment]
@alexarchambault I'm worried that the point of this feature request was misunderstood, so let me try to clarify.

What @mandrean is asking isn't so much "fetching all the dependencies of the `dependencyManagement` section.

The request is that the resolution should take `dependencyManagement` into account, in particular the dependency management included via the import scope, aka the BOM pattern. And that the equivalent capability be exposed via CLI:
* command line option to contribute an entry to `ResolutionParams.addForceVersion`, which IIUC is the equivalent of `dependencyManagement` in `pom.xml`
* command line option that takes a groupId:artifactId:version coordinate of a POM, then read its `dependencyManagement` and put those into `ResolutionParams.addForceVersion` recursively.

I'm in the same boat as @mandrean in that I'm blocked by the aforementioned bazel feature. Hope this helps to prioritize this issue accordingly.

**Comment by lefou** [comment]
As Mill also uses coursier, we also face the issue that users want to feed some BOM and expect to get the right versions selected/forced. Is that even possible with current coursier?

**Comment by rdesgroppes** [comment]
That would be extremely useful to build systems leveraging coursier, s.a. sbt, mill and bazel.

**Comment by nvachhar** [comment]
Is there a philosophical or technical reason to not implement BOM support? Or is this issue in search of someone to write the code?

**Comment by lefou** [comment]
> Resolution does take into account parent POMs / import scope during resolution though.

@alexarchambault Can you please clarify what you mean by that? Is coursier already reading and interpreting BOMs (with import scope)?

**Comment by ramzimaalej** [comment]
Any update on this issue?

**Comment by ramzimaalej** [comment]
@alexarchambault Can you point me to where I need to make the changes to add support for BOM?

**Comment by joan38** [comment]
I'd be happy to help too. We really need this feature.

**Comment by alexarchambault** [comment]
Just to clarify my understanding: I don't know if it makes sense to just resolve a BOM, like in
```text
$ coursier resolve com.fasterxml.jackson:jackson-bom:2.9.10
```

as `com.fasterxml.jackson:jackson-bom:2.9.10` isn't meant to be used as a simple dependency, nor to pull a class path.

But support for using BOMs while resolving other dependencies could be added, like
```text
$ coursier resolve com.pany:something:1.2.3 --bom com.fasterxml.jackson:jackson-bom:2.9.10 
```

That command would fetch the POM of `com.fasterxml.jackson:jackson-bom:2.9.10` (and integrate its parent POMs if needed), look at its `<dependencyManagement>` section, and use it to force some dependency versions (just like how `-E org:name:X.Y.Z` forces versions).

**Comment by alexarchambault** [comment]
Implementation-wise, going from the BOM dependency (like `com.fasterxml.jackson:jackson-bom:2.9.10`) to a `coursier.core.Project` that we'll read the `dependencyManagement` field of, should be the most tricky part. It's not only a matter of parsing the POM (with the help `coursier.maven.Pom`), parent POMs need to be folded into it, properties need to be substituted, etc. These concerns are handled from `coursier.core.Resolution` (by the `withDependencyManagement` method among others), but have not been written to be used in isolation from the resolution itself.

Another question is when should the versions be forced. At first, doing this from the `cli` module seemed the right thing to do, but that would restrict that feature to the CLI of coursier. As @lefou [suggests](https://github.com/coursier/coursier/issues/1390#issuecomment-732263031), this could also be useful from Mill, so this should be available from the API of coursier rather than only its CLI. That means support for this needs to be added in `coursier.Resolve`.

**Comment by alexarchambault** [comment]
> Can you please clarify what you mean by that? Is coursier already reading and interpreting BOMs (with import scope)?

@lefou Yes, coursier already takes into account the `<dependencyManagement>` section of POM files, to get the actual versions of dependencies during resolution in particular. And it takes into account parent POMs and import scope.

coursier just doesn't the possibility to its own users to manually pass their own BOM to be used during a resolution.

**Comment by alexarchambault** [comment]
> Adding a BOM-type dependency to the dependencies for resolution is currently not possible.

Why would one want to do that? BOMs aren't dependencies per se. They could be taken into account during resolution via specific options (like I suggest in https://github.com/coursier/coursier/issues/1390#issuecomment-1219224964), but they can't be handled like standard dependencies.

**Comment by lefou** [comment]
> > Adding a BOM-type dependency to the dependencies for resolution is currently not possible.
> 
> Why would one want to do that? BOMs aren't dependencies per se. They could be taken into account during resolution via specific options (like I suggest in [#1390 (comment)](https://github.com/coursier/coursier/issues/1390#issuecomment-1219224964)), but they can't be handled like standard dependencies.

we want to add BOMs in the sense of `<dependencyManagement>`. But there is currently no such mechanism in Mill or sbt. But in Mill, we want to add something like that: https://github.com/com-lihaoyi/mill/issues/1975, It's as you suggested for the CLI, but rather it needs to be in the API.

Motivation is the same as for BOMs in Maven: When working in projects with many dependencies and upstream provides BOMs (e.g. Vaadin, Maven) we want to include these BOMs in the dependencies (or some new kind of management dependencies) in the hope, that all explicit dependencies are aligned to the versions (and potential exclusions) declared in these BOMs.

**Comment by lefou** [comment]
About the API, i think there is currently no way to explicitly specify managed dependencies, there is only some way to map dependencies, but (mis-)using this mechanism to manage dependencies (or their versions) is rather tedious and makes not much sense, as coursier already supports dependency management when resolving transitive dependencies. It looks to me, the only missing part is currently that part of the API.

**Comment by aiuto** [comment]
Since Bazel and rules_jvm_external were mentioned, I thought I would chime in on how the Bazel team is thinking about SBOMs and what are short term plans are for the next few quarters.

A few years ago we started a project to replace the hard coded license handling in Bazel with one built in the extension language, Starlark. Using it, we can build up a list of the packages and their licenses used in an artifact.  To be more precise about that, we can, for any given artifact in a large repository, find the meta information about all the sources which are contained in that artifact, and then use that meta information to provide a list of license texts, license types, package names, versions, etc.

Because almost all Bazel users are multi-lingual we are taking a generic approach to package metadata, rather than building tools to work with specific types, such as Maven POMs. When we import a dependency, we convert global repository specific types (POM files) into a generic metadata declaration. Then we provide tools that can query those to provide license audits, SBOMs, or other kinds of build time tests and reporting.
 
Most code lives at https://github.com/bazelbuild/rules_license, but it will eventually thread through every repository downloader in the Bazel ecosystem, including rules_jvm_external.

**Comment by joan38** [comment]
> > Adding a BOM-type dependency to the dependencies for resolution is currently not possible.
> 
> Why would one want to do that? BOMs aren't dependencies per se. They could be taken into account during resolution via specific options (like I suggest in https://github.com/coursier/coursier/issues/1390#issuecomment-1219224964), but they can't be handled like standard dependencies.

For example in my company the Java team maintains an internal BOM and they don't like that we are using a build tool that does not read that file.
We have to manually override transitive deps from time to time for example when that log4j  security issue showed up.

**Comment by aiuto** [comment]
> We have to manually override transitive deps from time to time for example when that log4j security issue showed up.

That is a specific design point I am considering for what we are doing in Bazel. There are two aspects to this.

1. Users must be able to provide a mechanism to map any download URL to an alternate place, so that local repositories can assuredly provide local replacements for any transitive dependency.  Besides allowing a local patch to be quickly interposed (the log4j case), this is useful for building fully hermetic builds, or simply local caching.
2. Users will be able to provide a global hook to apply after downloading the package but before we start using it. That hook will be able to re-write package metadata based on local considerations.  The impetus from our end is that package authors are not always correct in their license declarations.  They might say something is Apache2, but then have a block of GPL code they neglected to add to their LICENSE.txt file. Users need a place where they could, if they desired, scan the entire import tree and deduce more accurate information.

**Comment by ramzimaalej** [comment]
Where do we stand on this? Are there any plans to fix this problem before the end of the year?

**Comment by joan38** [comment]
Sounds unlikely before the end of the year.
Maybe the Scala Center can help here? Cc @julienrf

**Comment by ramzimaalej** [comment]
If someone can provide some direction, I can probably help get it done.

**Comment by julienrf** [comment]
Thank you for this discussion.

@alexarchambault, does this comment correctly summarize the situation? https://github.com/coursier/coursier/issues/1390#issuecomment-1219291711

> A BOM-type POM file is one of packaging type `pom` and containing a `<dependencyManagement>` section.
> 
> Adding a BOM-type dependency to the dependencies for resolution is currently not possible.
> 
> But transitive BOM-type dependencies (that are added to the `<dependencyManagement><dependencies>` section with `<scope>import</scope>`) are already supported.

@alexarchambault, in the light of https://github.com/coursier/coursier/issues/1390#issuecomment-1219327644 and https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#bill-of-materials-bom-poms, do you agree that supporting BOM-type poms would be a good thing to have in Coursier?

**Comment by rdesgroppes** [comment]
To give a concrete case where this feature would be more than desirable is with distributed computing frameworks (Apache Beam/Google Cloud Dataflow, Apache Flink, Apache Spark, etc.), where we frequently encounter the following pattern: container images (vanilla or custom) are deployed with a predefined and mostly consistent set of dependencies whose versions are finely selected to be compatible with each other. Their selection is sometimes the result of considerable effort (example: [BOM: com.google.cloud:libraries-bom:25.2.0](https://storage.googleapis.com/cloud-opensource-java-dashboard/com.google.cloud/libraries-bom/25.2.0/artifact_details.html)) and, when at least one of them is incremented, all direct dependencies are re-examined with the same care in order to increment them if necessary, and so on, until a new version of the container image is consistent as a whole.
For each such container image, a BOM is produced that provides an exhaustive list of the selected dependencies (either `provided` when shared with the framework's runtime and/or to save assembly and upload times, or `compile` otherwise). This BOM is consumed by users in order to compile, test and assemble "fat" JARs (or uber-JARs) containing their application code as well as the additional dependencies they need. It is therefore the BOM that sets the pace for the versions guaranteed to be available in the deployment and consistent with each other.
What's missing is therefore a mechanism that would allow to automatically pick the individual dependency versions from these BOM.

**Comment by alexarchambault** [comment]
Alright, so just to clarify my position on this: I agree accepting BOMs coordinates (and maybe also a path to a BOM file) would be a good thing. That feature needs to be implemented in the API (so that Mill can re-use it), and an option for it, such as `--bom`, needs to be added to the CLI.

I gave a few pointers in https://github.com/coursier/coursier/issues/1390#issuecomment-1219238978 about the implementation. Basically, a field needs to be added to `coursier.Resolve`, typed as… a custom ADT maybe, that could hold either coordinates (org:name:version), or a path to a file (only on the JVM, not when targetting Scala.js).

Then right before actually running the resolution (at the beginning of `Resolution.ioWithConflicts0` I guess), we should get the BOM content, parse it, and add its entries to `Resolution#resolutionParams` via `addForceVersion`.

**Comment by alexarchambault** [comment]
The most tricky part should be the few refactorings to do in `coursier.core.Resolution`, so that one can get the BOM content as a `coursier.core.Project` instance (in order to use its `dependencyManagement` field).

**Comment by alexarchambault** [comment]
About tests, I would add some API tests in [`ResolveTests`](https://github.com/coursier/coursier/blob/67daad6a9786c0f9187617ef555e8ee5b9039a06/modules/coursier/shared/src/test/scala/coursier/ResolveTests.scala), and maybe a CLI test in [`coursier.clitests.ResolveTests`](https://github.com/coursier/coursier/blob/67daad6a9786c0f9187617ef555e8ee5b9039a06/modules/cli-tests/src/main/scala/coursier/clitests/ResolveTests.scala).

**Comment by eed3si9n** [comment]
When people say support BOM don't they usually mean that they can omit the version number in the dependency specification, like:

```xml
<dependency>
  <groupId>junit</groupId>
  <artifactId>junit</artifactId>
  <scope>test</scope>
</dependency>
```

and the exact version would be determined by the parent POM or Bill-of-Material (BOM)?

**Comment by alexarchambault** [comment]
> When people say support BOM don't they usually mean that they can omit the version number in the dependency specification

I guess this can be done in a second time. Maybe… we could accept versions like `_`, and replace them with whatever is in the BOM (but then this would fail at runtime if there's no BOM, or for modules not in the BOM). But this would work for the CLI for sure.

**Comment by julienrf** [comment]
@ramzimaalej, are you still interested in implementing this feature? Do you need any additional information?

**Comment by lefou** [comment]
Because I have the impression, that different persons may expect different things from this issue, I'd like to outline what exactly I envision. Please correct me, if you see things different or if you think, I should place parts of it into a separate issue, thanks.

The different between Maven and its dependency management library Aether and this very project Coursier is, that Maven is also the build tool (a package provider), whereas Coursier is typically only used to resolve and fetch dependencies (a package consumer). 

A package consumer:
* resolved the dependencies and all transitive dependencies (`<dependencies/>`)
* infer additional rules and applies them to resolve conflicts or decide between artifacts (`<dependencyManagement/>`, `<exclusion/>`)

A package provider:
* is a package consumer
* additionally provides rules for downstream consumers (`<dependencyManagement/>`)

Coursier does the package consumer part very well already. But it lacks support for the provider part. In fact, artifact publishing in Mill completely bypasses Coursier, and I think sbt does it too. Now, we aim for making both aspects more consistent.

Although, we are able to generate arbitrary complex `pom.xml` (without letting Coursier know), what we really want is make Coursier able to understand what we would write into that `pom.xml`. That means, if we add a `<dependencyManagement/>` section to the `pom.xml` which specified an exact version for a dependency, we expect that Coursier is also using this version. And when we add a BOM dependency with an `import`-scope to that section, we expect that Coursier also reads this BOM and takes all management rules into account, when resolving the transitive dependency set.

Again, the idea is to make both aspects consistent: the way Coursier resolves dependencies, and the way, how we persist our dependency rules downstream via `pom.xml` files.

Hence, we need an API to tell Coursier how we are going to rule downstream package users via dependency management. The Coursier Resolve API needs to accept a new parameter: A List or Set of dependencies which are used for dependency management. These either are normal dependencies, e.g. to declare or force a version, or `import`-scope dependencies, which means they refer to a pom-type artifact which needs to be fetched and which contains more dependency management entries, which should be considered as resolution rules.

**Comment by caeus** [comment]
Any help needed?

**Comment by lefou** [comment]
> Any help needed?

Yes. Alex already made some comments (https://github.com/coursier/coursier/issues/1390#issuecomment-1282409459 and following) what kind of changes and where these changes might be needed.

**Comment by alexarchambault** [comment]
Should be fixed by https://github.com/coursier/coursier/pull/3097 and https://github.com/coursier/coursier/pull/3143. These will be part of the upcoming `2.1.17` release.

**Comment by joan38** [comment]
Thanks for this, is there any support planned in the scala-cli?

**Comment by tgodzik** [comment]
I think once we update coursier in scala-cli this will be supported. @alexarchambault ?
